modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java

```diff
@@ -1557,41 +1598,67 @@ public DDMStructure updateStructure(
 		DDMStructure structure = ddmStructurePersistence.findByPrimaryKey(
 			structureId);
 
-		User user = _userLocalService.getUser(userId);
+		return _updateStructure(
+			userId, structureId, parentStructureId, structureKey, nameMap,
+			descriptionMap, definition, structure, serviceContext);
+	}
 
-		structure.setUserId(userId);
-		structure.setParentStructureId(parentStructureId);
+	@Indexable(type = IndexableType.REINDEX)
+	@Override
+	public DDMStructure updateStructure(
+			String externalReferenceCode, long userId, long groupId,
+			long structureId, long parentStructureId, long classNameId,
+			Map<Locale, String> nameMap, Map<Locale, String> descriptionMap,
+			DDMForm ddmForm, DDMFormLayout ddmFormLayout,
+			ServiceContext serviceContext)
+		throws PortalException {
 
-		if (Validator.isNotNull(structureKey)) {
-			structureKey = StringUtil.toUpperCase(structureKey.trim());
+		DDMStructure structure = ddmStructurePersistence.findByPrimaryKey(
+			structureId);
 
-			_validateStructureKey(
-				structureId, structure.getGroupId(), structure.getClassNameId(),
-				structureKey);
+		if (externalReferenceCode != null) {
```

Comment:
I think it's better to encapsulate this logic inside the `_validateExternalReferenceCode` method, so you would end up with something like this [example](https://github.com/liferay/liferay-portal/blob/12d419627dec40e879aeec074fd38940938ae694/modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectActionLocalServiceImpl.java#L511).
FF verification can also be included within the validation method.

Reply:
Sure, will do!